src/ngx_base_fetch.cc

```diff
@@ -254,7 +255,12 @@ ngx_int_t NgxBaseFetch::CopyBufferToNginx(ngx_chain_t** link_ptr) {
 ngx_int_t NgxBaseFetch::CollectAccumulatedWrites(ngx_chain_t** link_ptr) {
   ngx_int_t rc;
   Lock();
-  rc = CopyBufferToNginx(link_ptr);
+  if (successful_ || response_headers()->IsErrorStatus()) {
+    rc = CopyBufferToNginx(link_ptr);
+  } else {
+    // See comment in CollectHeaders.
+    rc = NGX_OK;
```

Comment:
Out of curiosity, did you try returning `NGX_ERROR` here? That would make the caller release (and detach from) the base fetch right away, and return `NGX_HTTP_INTERNAL_SERVER_ERROR`. `NgxBaseFetch::ReadCallback` would then discard further events associated to this detached based fetch (a tiny bit more efficient), and we'll call `ngx_log_error` about this happening, which may have some diagnostic value?
(This would also apply to `CollectHeaders` below).